gateserver/controller_protocol.py

```diff
@@ -0,0 +1,103 @@
+
+See https://github.com/fmfi-svt-gate/server/wiki/Controller-%E2%86%94-Server-Protocol .
+"""
+
+from .utils.structparse import mystruct, t
+from enum import Enum
+import nacl.raw as nacl
+
+PROTOCOL_VERSION = bytes([0x00,0x01])
+
+class BadMessageError(Exception): pass
+
+def checkmsg(expression, errmsg):
+    if not expression: raise BadMessageError(errmsg)
+
+class MsgType(Enum):
+    OPEN = 1
+
+class ResponseStatus(Enum):
+    OK        = 0x01
+    ERR       = 0x10
+    TRY_AGAIN = 0x11
+
+Packet = mystruct('Packet',
+                  (t.bytes(2) , 'protocol_version'),
+                  (t.bytes(6) , 'controller_id'   ),
+                  (t.bytes(18), 'nonce'           ))
+
+Request = mystruct('Request',
+                   (t.uint8, 'msg_type'))
+
+Response = mystruct('Response',
+                    (t.uint8, 'msg_type'),
+                    (t.uint8, 'status'  ))
+
+def crypto_unwrap(packet, key, payload):
+    return nacl.crypto_secretbox_open(payload,
+        packet.controller_id + packet.nonce, key)
+
+def crypto_wrap(packet, key, payload):
+    return nacl.crypto_secretbox(payload,
+        packet.controller_id + packet.nonce, key)
+
+def parse_packet(buf):
+    """Parses the packet into a `Packet` struct with the body in `tail`."""
+    packet = Packet.unpack_with_tail(buf)
+    checkmsg(packet.protocol_version == PROTOCOL_VERSION,
+        'Invalid protocol version')
+    return packet
+
+def payload_from_packet(payload_struct, packet, key):
+    """Decrypts the packet payload and parses the request/response header.
+
+    Returns the parsed payload (as struct), with the `msg_type_e` property set
+    to the parsed `MsgType`.
+    """
+    assert payload_struct in [Request, Response]
+    try:
+        payload = crypto_unwrap(packet, key, packet.tail)
+    except ValueError as e:
+        raise BadMessageError('Decryption failed') from e
+    parsed = payload_struct.unpack_with_tail(payload)
+    try:
+        parsed.msg_type_e = MsgType(parsed.msg_type)
+    except ValueError as e:
+        raise BadMessageError('Unknown message type') from e
+    return parsed
+
+def request_from_packet(packet, key):
+    """Decrypts the request packet payload and parses the request.
+
+    Returns the parsed Request, with the `msg_type_e` property set
+    to the parsed `MsgType`.
+    """
+    return payload_from_packet(Request, packet, key)
+
+def response_from_packet(packet, key):
+    """Decrypts the response packet payload and parses the response.
+
+    Returns the parsed Response, with the `msg_type_e` and `status_e` properties
+    set to the parsed `MsgType` and `ResponseStatus`, respectively.
+    """
+    response = payload_from_packet(Response, packet, key)
+    try:
+        response.status_e = ResponseStatus(response.status)
+    except ValueError:
+        raise BadMessageError('Unknown status {}'.format(response.status))
+    return response
+
+def response_packet_for(request_packet, response, key):
```

Comment:
Zvaz premenovat na `packet_from_response`, pre konzistenciu s ostatnymi funkciami. (Ale je mi to jedno.)